rfcs/0008-distributed-dns.md

```diff
@@ -0,0 +1,241 @@
+- `MultiClusterFailureThreshold`
+  - When absent defaults to `66`, what percentage of clusters need to be reporting an endpoint unhealthy on a remote IP for it to be removed from the zone by a cluster that does not own that IP.
+- `clusterID`
+  - When absent is generated in code (more information below). A unique string to identify this cluster apart from all other clusters.
+
+New fields added to the DNS Record CRD:
+- `HealthCheck`
+  - Propagated from the DNS Policy, to facilitate reconciling health checks in the DNS controller.
+
+### Changes to DNS Policy reconciler (kuadrant operator)
+
+This will primarily need to be made aware of the new fields on the DNS Policy (Healthcheck block) that it will need to propagate into any DNS Records that it creates or updates. Otherwise the logic of this controller remains largely the same.
+
+### DNS Operator
+
+This component takes the DNS Records output by the kuadrant-operator and ensure they are accurately reflected in the DNS Provider zone and currently assumes the DNS Record is the source of truth and always ensures that the zone in the DNS Provider matches the local DNS Record CR.
+
+There are several changes required in this component:
+
+#### Generate a ClusterID
+
+The cluster will need a deterministic manner to generate an ID that is unique enough to not clash with other clusterIDs, but determinstic enough that it can be regenerated if ever required.
+
+This clusterID is used for multiple purposes:
+- To identify which A/CNAME records were created by which cluster
+- To see how many clusters have interacted with this zone (number of unique cluster IDs in all A/CNAME records)
+- To report a failing health check
+- To count how many clusters are reporting a failing health check.
+
+The suggestion [here](https://groups.google.com/g/kubernetes-sig-architecture/c/mVGobfD4TpY/m/nkdbkX1iBwAJ) is to use the UID of the `kube-system` namespace, as this [cannot be deleted easily](https://groups.google.com/g/kubernetes-sig-architecture/c/mVGobfD4TpY/m/lR4SnaqpAAAJ) - so is practically speaking indelible.
+
+We can take the UID of that namespace and apply a hashing algorithm to result in 6-7 character code which can be used as the clusterID for the local cluster.
+
+#### Update GUID Logic
+
+Currently the DNS Operator builds the LB CNAME with a guid based on the name of the gateway, as this could potentially be different on different clusters, this will need to be updated to generate based on the zone's root hostname, to ensure it is consistently generated on all clusters that interact with this zone.
+
+#### Build DNS Records from the provider zone
+
+When the DNS Operator reconciles a kuadrantRecord and will first pull the most recent records from the DNS Provider and construct a separate zoneRecord.
```

Comment:
Is this "kuadrantRecord" just a DNSRecord, or a new resource?

Reply:
yeah I found this terminology confusing also. But it is just another DNSRecord

Reply:
yeah, I was just trying to simplify the 2 different concepts (zoneRecord is a local cache of the values in the zone, kuadrantZone is the local description of what Kuadrant wants to be present in the zone). Calling them by descriptions like "the record created by Kuadrant" and "the record that reflects the zone" was making sentence very long and confusing.